torch/csrc/Exceptions.cpp

```diff
@@ -137,5 +137,89 @@ ValueError::ValueError(const char *format, ...) {
   va_end(fmt_args);
 }
 
+// ATen warning handler for Python
+using warning_buffer_t =
+  std::vector<std::pair<c10::SourceLocation, std::string>>;
+
+static warning_buffer_t warning_buffer;
+// To avoid deadlocks, if both the GIL and this lock is needed,
+// The GIL needs to be acquired first.
+static std::mutex warning_buffer_mutex;
+
+static void py_warning_handler(
+    const c10::SourceLocation& source_location,
+    const std::string& msg) {
+  std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+  warning_buffer.push_back({source_location, msg});
+};
+
+EnforceWarningBuffer::EnforceWarningBuffer() noexcept(true): prev_handler(c10::Warning::get_warning_handler()) {
+  c10::Warning::set_warning_handler(&py_warning_handler);
+}
+
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+/// NOLINTNEXTLINE(bugprone-exception-escape)
+EnforceWarningBuffer::~EnforceWarningBuffer() noexcept(false) {
+  c10::Warning::set_warning_handler(prev_handler);
+
+  bool has_warnings;
+  {
+    std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+    has_warnings = warning_buffer.size() > 0;
```

Comment:
Hmmm. While it is certainly true that you avoid the deadlock this way, taking out a lock to test size, and then releasing the lock, and then taking the lock out later is definitely a concurrency antipattern. What if another thread modifies `warning_buffer` in-between the time you check and then when you actually run?

But this raises further questions in my mind: why is the warning buffer global (as opposed to something thread local?)

Reply:
Because it needs to be accessed from the static `at::warn()` function.
Is there a way to bind to bind to call `static void set_warning_handler(handler_t handler);` with a handler which is a function associated with this class instance? For example if I had a `EnforceWarningBuffer::warn_handler()` (non static function).

Reply:
It is not thread local mainly because that would force us to duplicate all the logic that currently exists for errors to move them from one thread to the other whenever multithreaded is used.

Reply:
Well, if the warning handler takes a straight up function pointer there's nothing you can do, but if you change it to accept something more interesting you would be in better shape. Like it takes a pointer with a virtual method on it to handle.

The justification for not TLS feels flimsy to me, though I understand from a local perspective why it would be nicer to avoid having to solve the TLS transfer across threads problem. In particular, the reason why it feels dodgy to me is because it means one thread can queue warnings, and then they magically show up from another thread when returning to Python. Warnings will literally escape the scopes they were supposed to live in...

Reply:
After discussion. I'll update the PR to detect when such "escape" can happen. And warn the user in that case.